kubetest2/pkg/metadata/writer_test.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+	"time"
+)
+
+// fake time.Now that always increments by one second
+func makeFakeNow() func() time.Time {
+	var t time.Time
+	return func() time.Time {
+		t = t.Add(time.Second)
+		return t
+	}
+}
+
+// junitError impl for testing
+type junitError struct {
+	name   string
+	stderr string
+	stdout string
+}
+
+// assert that junitError is actually a JUnitError
+var _ JUnitError = &junitError{}
+
+func (j junitError) Error() string {
+	return j.name
+}
+
+func (j *junitError) Stderr() string {
+	return j.stderr
+}
+
+func (j *junitError) Stdout() string {
+	return j.stdout
+}
+
+func TestWriter(t *testing.T) {
+	type step = struct {
+		name        string
+		doStep      func() error
+		expectError bool
+	}
+
+	var testCases = []struct {
+		name           string
+		steps          []step
+		expectedOutput string
+	}{
+		{
+			name: "all passing",
+			steps: []step{
+				{
+					name:   "noop",
+					doStep: func() error { return nil },
+				},
+			},
+			expectedOutput: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><testsuite failures=\"0\" tests=\"1\" time=\"3\">\n    <testcase name=\"noop\" time=\"1\"></testcase>\n</testsuite>",
+		},
+		{
+			name: "one failed step",
+			steps: []step{
+				{
+					name:        "always fails",
+					doStep:      func() error { return errors.New("oh noes") },
+					expectError: true,
+				},
+			},
+			expectedOutput: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><testsuite failures=\"1\" tests=\"1\" time=\"3\">\n    <testcase name=\"always fails\" time=\"1\">\n        <failure>oh noes</failure>\n    </testcase>\n</testsuite>",
+		},
+		{
+			name: "one failed step with junitError",
+			steps: []step{
+				{
+					name: "always fails (junitError)",
+					doStep: func() error {
+						return &junitError{
+							name:   "on noes",
+							stdout: "uh oh",
+							stderr: "whoops",
+						}
+					},
+					expectError: true,
+				},
+			},
+			expectedOutput: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><testsuite failures=\"1\" tests=\"1\" time=\"3\">\n    <testcase name=\"always fails (junitError)\" time=\"1\">\n        <failure>on noes</failure>\n        <system-out>uh oh</system-out>\n        <system-err>whoops</system-err>\n    </testcase>\n</testsuite>",
```

Comment:
add a testcase for multiple steps? also this output is kinda hard to read :-) we probably want new lines?

Reply:
sure I can reformat this into `backtick` strings
adding multiple steps

Reply:
thanks - I'm more worrying about the actual output will be a single line which will be hard to read :-)

Reply:
done